src/imp/run.c

```diff
@@ -0,0 +1,333 @@
+ *
+ *  - wait for child to terminate successfully
+ *
+ *  - execute command
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <stdio.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <errno.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <pwd.h>
+#include <signal.h>
+
+#include "src/libutil/kv.h"
+
+#include "imp_log.h"
+#include "imp_state.h"
+#include "impcmd.h"
+#include "privsep.h"
+
+extern char **environ;
+
+static const cf_t *imp_run_lookup (struct imp_state *imp,
+                                   const char *name)
+{
+    const char *path;
+    const cf_t *cf;
+    const cf_t *run = cf_get_in (imp->conf, "run");
+
+    /*  Check for [run.name] configuration table:
+     */
+    if (!run || !(cf = cf_get_in (run, name)))
+        imp_die (1, "run: %s: no configuration found", name);
+
+    /*  Check for required members of [run.name]:
+     *   - 'path' must exist and be an absolute path
+     */
+    if (!(path = cf_string (cf_get_in (cf, "path")))
+        || path[0] == '\0'
+        || path[0] != '/')
+        imp_die (1, "run: %s: path is missing or invalid", name);
+
+    return cf;
+}
+
+static bool run_user_allowed (const cf_t *cf_run)
+{
+    struct passwd *pwd;
+
+    if (!(pwd = getpwuid (getuid ())) || !pwd->pw_name)
+        imp_die (1, "Unable to lookup user");
+
+    return cf_array_contains (cf_get_in (cf_run, "allowed-users"),
+                              pwd->pw_name);
+}
+
+static bool run_env_var_allowed (const cf_t *allowed_env,
+                                 const char *name)
+{
+    if (strcmp (name, "FLUX_JOB_ID") == 0
+        || strcmp (name, "FLUX_JOB_USERID") == 0)
+        return true;
+    return cf_array_contains_match (allowed_env, name);
+}
+
+/*  Return a kv structure with the environment for this run command.
+ */
+static struct kv *get_run_env (struct kv *kv, const cf_t *allowed_env)
+{
+    struct kv *kv_env;
+    const char *var = NULL;
+
+    if (!(kv_env = kv_split (kv, "IMP_RUN_ENV_")))
+        return NULL;
+
+    while ((var = kv_next (kv_env, var))) {
+        if (!run_env_var_allowed (allowed_env, var))
+            kv_delete (kv_env, var);
+    }
+
+    /*  Capture uid that ran the imp as the "owner" */
+    if (kv_put (kv_env,
+                "FLUX_OWNER_USERID",
+                KV_INT64,
+                (int64_t) getuid()) < 0)
+        imp_die (1, "failed to put FLUX_IMP_USERID in environment");
+
+    return kv_env;
+}
+
+static void __attribute__((noreturn))
+imp_run (const char *name,
+         const cf_t *run_cf,
+         struct kv *kv_env)
+{
+    const char *path;
+    struct passwd *pwd;
+    char **env;
+    const char *args[2];
+    int exit_code;
+
+    if (!(path = cf_string (cf_get_in (run_cf, "path")))
+        || path[0] != '/')
+        imp_die (1, "run: %s: invalid path", name);
+
+    if (path[0] != '/')
+        imp_die (1, "run %s: relative path not allowed", name);
```

Comment:
Wondering if any other checks are needed here, such as verifying that the executable or the path containing it are not writable by others?

Reply:
Yeah, good thought. Too bad the functions to check file and directory ownership and permissions are private to `libutil/cf.c` -- that wasn't very foresighted. :disappointed: 

We'll have to maybe split out those functions in libutil...